### PR TITLE
FIX: fix list closed orders api limit

### PR DESCRIPTION
**Pull request by kbearXD**

### CODE REVIEW

File: pkg/exchange/max/exchange.go [hunk 1]
```diff
@@ -325,11 +341,21 @@ func (e *Exchange) queryClosedOrdersByLastOrderID(
...
 }
 
-func (e *Exchange) queryClosedOrdersByTime(ctx context.Context, symbol string, since, until time.Time) (orders []types.Order, err error) {
+func (e *Exchange) queryClosedOrdersByTime(ctx context.Context, symbol string, since, until time.Time, orderByType maxapi.OrderByType, limit uint) (orders []types.Order, err error) {
 	if err := e.closedOrderQueryLimiter.Wait(ctx); err != nil {
 		return orders, err
 	}
 
+	// there is since limit for closed orders API
+	sinceLimit := time.Date(2018, time.January, 1, 0, 0, 0, 0, time.Local)
```
Comment (ycdesu) [inline pkg/exchange/max/exchange.go:350]: This time can be moved to package level to be shared with other functions.

**Review by ycdesu — APPROVED** [review]

File: pkg/exchange/max/exchange.go [hunk 2]
```diff
@@ -270,18 +270,34 @@ func (e *Exchange) QueryOpenOrders(ctx context.Context, symbol string) (orders [
...
-	log.Warn("!!!MAX EXCHANGE API NOTICE!!! the since/until conditions will not be effected on closed orders query, max exchange does not support time-range-based query")
 	if !since.IsZero() || !until.IsZero() {
-		return e.queryClosedOrdersByTime(ctx, symbol, since, until)
+		return e.queryClosedOrdersByTime(ctx, symbol, since, until, maxapi.OrderByAsc, 1000)
 	}
 
 	return e.queryClosedOrdersByLastOrderID(ctx, symbol, lastOrderID)
 }
 
+func (e *Exchange) QueryClosedOrdersDesc(
```
Comment (bailantaotao) [inline pkg/exchange/max/exchange.go:283]: 沒有人呼叫他？

**Review by bailantaotao — APPROVED** [review]

File: pkg/exchange/max/exchange.go (hunk 2 above)
Comment (kbearXD) [reply pkg/exchange/max/exchange.go:283]: dca2 會用到

File: pkg/exchange/max/exchange.go (hunk 1 above)
Comment (kbearXD) [reply pkg/exchange/max/exchange.go:350]: will use getLaunchDate

File: pkg/exchange/max/exchange.go [hunk 3]
```diff
@@ -270,18 +270,34 @@ func (e *Exchange) QueryOpenOrders(ctx context.Context, symbol string) (orders [
...
+	ctx context.Context, symbol string, since, until time.Time, lastOrderID uint64,
```
Comment (c9s) [inline pkg/exchange/max/exchange.go:284]: lastOrderID should be optional, not all exchanges support this

File: pkg/exchange/max/exchange.go [hunk 4]
```diff
@@ -270,18 +270,34 @@ func (e *Exchange) QueryOpenOrders(ctx context.Context, symbol string) (orders [
...
+) ([]types.Order, error) {
+	closedOrders, err := e.queryClosedOrdersByTime(ctx, symbol, since, until, maxapi.OrderByDesc, 1000)
+	if lastOrderID == 0 {
+		return closedOrders, err
+	}
+
+	var filterClosedOrders []types.Order
+	for _, closedOrder := range filterClosedOrders {
+		if closedOrder.OrderID > lastOrderID {
```
Comment (c9s) [inline pkg/exchange/max/exchange.go:293]: should be compared only when lastOrderID is not zero

File: pkg/exchange/max/exchange.go [hunk 5]
```diff
@@ -270,18 +270,34 @@ func (e *Exchange) QueryOpenOrders(ctx context.Context, symbol string) (orders [
...
+func (e *Exchange) QueryClosedOrdersDesc(
+	ctx context.Context, symbol string, since, until time.Time, lastOrderID uint64,
+) ([]types.Order, error) {
+	closedOrders, err := e.queryClosedOrdersByTime(ctx, symbol, since, until, maxapi.OrderByDesc, 1000)
+	if lastOrderID == 0 {
+		return closedOrders, err
+	}
+
+	var filterClosedOrders []types.Order
+	for _, closedOrder := range filterClosedOrders {
```
Comment (c9s) [inline pkg/exchange/max/exchange.go:292]: you're iterating an empty array?

File: pkg/exchange/max/exchange.go [hunk 6]
```diff
@@ -325,11 +341,24 @@ func (e *Exchange) queryClosedOrdersByLastOrderID(
 	return types.SortOrdersAscending(orders), nil
 }
 
-func (e *Exchange) queryClosedOrdersByTime(ctx context.Context, symbol string, since, until time.Time) (orders []types.Order, err error) {
+func (e *Exchange) queryClosedOrdersByTime(ctx context.Context, symbol string, since, until time.Time, orderByType maxapi.OrderByType, limit uint) (orders []types.Order, err error) {
```
Comment (c9s) [inline pkg/exchange/max/exchange.go:344]: we can simply use the default limit

Comment (c9s) [reply pkg/exchange/max/exchange.go:344]: by the maximum number

File: pkg/exchange/max/exchange.go (hunk 4 above)
Comment (kbearXD) [reply pkg/exchange/max/exchange.go:293]: Yes, check lastOrderID at line 287.

File: pkg/exchange/max/exchange.go (hunk 5 above)
Comment (kbearXD) [reply pkg/exchange/max/exchange.go:292]: You are right.....

File: pkg/exchange/max/exchange.go (hunk 3 above)
Comment (kbearXD) [reply pkg/exchange/max/exchange.go:284]: I think we can think if lastOrderID is zero means we don't need to care it ?

File: pkg/exchange/max/exchange.go (hunk 6 above)
Comment (kbearXD) [reply pkg/exchange/max/exchange.go:344]: ok!

**Review by c9s — APPROVED** [review]

**Pull request merged**